### PR TITLE
Added beginning of i18n.

### PR DESCRIPTION
**Pull request by zond**
@tttppp, @JorenC, anyone else interested.

This is a branch with the beginnings of i18n support in dipact.

It works like this:

1) app.js (that starts the app) uses `(window.navigator.userLanguage || window.navigator.language || "en").split(/-/)[0]` to set the global variable `window.Globals.lang`.
2) The new file `/static/js/langs.js` contains a big Object looking like (right now):
```
const langs = {
	en: {
		"A digital version of the classic game of Diplomacy. Sign in to play.":
			"A digital version of the classic game of Diplomacy. Sign in to play.",
		"Stay logged in": "Stay logged in",
		"Sign in with Google": "Sign in with Google",
		"{0} (You)": "{0} (You)",
		"Doesn't match the order for {0}": "Doesn't match the order for {0}",
		Austria: "Austria",
		France: "France",
		Germany: "Germany",
		Italy: "Italy",
		Turkey: "Turkey",
		Russia: "Russia",
		England: "England",
	},
};
```
3) The new global function (defined on `window`, so available _everywhere_) `tr(s)` returns the translation of `s` in the part of `locales` defined by `window.Globals.lang`. I.e. `tr('Stay logged in')` will return `Stay logged in` if the browser wants the language `en`. If the language in `window.Globals.lang` isn't defined, it will return whatever's defined for `en`. If the translation for the string itself isn't defined, the same string will be returned after adding it to `window.Globals.missingTranslations`.

4) I added a function to _all_ strings in the app called `format`. It replaces occurences of `{[any number]}` in the string with argument nr [any number] among the arguments. I.e. calling `"Hello {0} what's up?".format("Martin")` will return `"Hello Martin what's up?"`.

5) I started replacing a few strings in the app using the new `tr` function as an example.

Do you think this is a good approach?

I'm very interested in everyones opinion since I don't have the time and energy to do all of this on my own. I hope others will help with this work. Translators can add new languages in `langs.js`, other enthusiasts can replace text in the rest of the app with `tr(text)` according to the pattern in this pull request.

### CODE REVIEW

**Comment by tttppp** [comment]
I think this sounds like a good approach for all the static strings. I'm nervous about using it for e.g. the nation names or other data coming from the backend.
1. Nation names, variant names, etc. are all used as ids, so there's a (very) small risk we'll send back the translation by mistake
2. The map won't be translated, so we'll still have some English displayed
3. This feels like a blunt tool, and we may run into issues if a nation can be translated two ways (e.g. maybe France called itself different things at different points in history?)

File: resources/js/app.js [hunk 1]
```diff
@@ -16,7 +17,31 @@ if (fakeID) {
 	serverURL.searchParams.set("fake-id", fakeID);
 }
 
+String.prototype.format = function () {
+	var formatted = this;
+	for (var i = 0; i < arguments.length; i++) {
+		var regexp = new RegExp("\\{" + i + "\\}", "gi");
+		formatted = formatted.replace(regexp, arguments[i]);
```
Comment (tttppp) [inline resources/js/app.js:24]: Is order preserved when iterating through a map in JS?  Due to the way string replacement happens we may want to ensure the replacements happen in a particular order. e.g. We might want to replace "English Channel" before we replace "English".

**Comment by tttppp** [comment]
I'm sure you already know this, but some languages require right-to-left text, or changes to whitespace/punctuation, which this approach might struggle with. However I think it looks like a great approach to help a large number of users.

File: resources/js/app.js (hunk 1 above)
Comment (zond) [reply resources/js/app.js:24]: The arguments is an array, right? So it should just replace `{0}` with the first argument, etc.

**Comment by zond** [comment]
> I'm sure you already know this, but some languages require right-to-left text, or changes to whitespace/punctuation, which this approach might struggle with. However I think it looks like a great approach to help a large number of users.

Well, if you add unicode to reverse the text in the translation we can reverse it, and the punctuation should be a part of the translations, so it should work, no?

**Comment by zond** [comment]
> I think this sounds like a good approach for all the static strings. I'm nervous about using it for e.g. the nation names or other data coming from the backend.

I _think_ it's fine. If it's not translated, the output will be the input string, so the server sending new/changed stuff should be OK.

> 1. Nation names, variant names, etc. are all used as ids, so there's a (very) small risk we'll send back the translation by mistake

Yeah, we must make sure that nothing sent to the server is translated - only things displayed to the user.

> 2. The map won't be translated, so we'll still have some English displayed

True, but that's OK I think - I consider translating the maps out of scope.

> 3. This feels like a blunt tool, and we may run into issues if a nation can be translated two ways (e.g. maybe France called itself different things at different points in history?)

Yeah, but unless we want to translate everything super early in the pipeline (variant and map definitions) that will also be out of scope. I find it acceptable if two variants that in English would have `France` and `France` as names for France, but in French would have `[something french]` and `[something else in french]` as names still have the same name even in French. We can blame the fact that most variants are not our projects, and only exist in one language, and we don't want to manage translating projects that aren't ours.

**Comment by tttppp** [comment]
> > I'm sure you already know this, but some languages require right-to-left text, or changes to whitespace/punctuation, which this approach might struggle with. However I think it looks like a great approach to help a large number of users.
> 
> Well, if you add unicode to reverse the text in the translation we can reverse it, and the punctuation should be a part of the translations, so it should work, no?
I was thinking code like this won't work in all languages:
```tr("{0} (You)").format(tr(nation))```

**Review by tttppp — APPROVED** [review]
I think this is a big improvement, and we can always update it if we ever added multilanguage support to the backend.

**Comment by tttppp** [comment]
> > > I'm sure you already know this, but some languages require right-to-left text, or changes to whitespace/punctuation, which this approach might struggle with. However I think it looks like a great approach to help a large number of users.
> > 
> > 
> > Well, if you add unicode to reverse the text in the translation we can reverse it, and the punctuation should be a part of the translations, so it should work, no?
> 
> I was thinking code like this won't work in all languages:
> `tr("{0} (You)").format(tr(nation))`

Maybe it will actually. Tokenised replacement is pretty powerful.

**Comment by zond** [comment]
> > > I'm sure you already know this, but some languages require right-to-left text, or changes to whitespace/punctuation, which this approach might struggle with. However I think it looks like a great approach to help a large number of users.
> > 
> > 
> > Well, if you add unicode to reverse the text in the translation we can reverse it, and the punctuation should be a part of the translations, so it should work, no?
> 
> I was thinking code like this won't work in all languages:
> `tr("{0} (You)").format(tr(nation))`

I'm thinking that some languages have `England (You)` expressed as `[word for England] ([word for you])`, those are the easiest cases.
Other languages have `([word for you]) [word for England]`. In those languages `tr` would return `(You) {0}`.
The worst ones have `([uoy rof drow]) [dnalgnE rof drow]`. In those languages `tr` would return `[utf for right-to-left]{0} ([word for you])`.

Does that make sense?

**Comment by zond** [comment]
@tttppp In any case, I don't want to merge this now - I see this as the 'translation branch'. Let all translation work merge into this branch, and when we feel that it's complete enough not to be embarrassing we merge it to master?

**Comment by zond** [comment]
Closing this since this is based on the old version and the new standard react version is probably the way to go.

**Pull request closed without merging**